eigen_cpp/lib/sparse_solver.cpp

```diff
@@ -0,0 +1,86 @@
+#include "sparse_solver.h"
+
+bool report_error(Eigen::ComputationInfo info) {
+    switch(info) {
+    case NumericalIssue:
+      std::cerr << "Solver Failed: Numerical Issue (prerequisites not met)" << std::endl;
+      return true;
+    case NoConvergence:
+      std::cerr << "Solver Failed: Did not converge" << std::endl;
+      return true;
+    case InvalidInput:
+      std::cerr << "Solver Failed: Invalid input" << std::endl;
+      return true;
+    case Success:
+    default:
+      return false;
+    }
+  }
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+spmat& SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::c_to_eigen(c_spmat* ptr)
+{
+  return *reinterpret_cast<spmat*>(ptr);
+}
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+c_spmat* SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::eigen_to_c(spmat& ref)
+{
+  return reinterpret_cast<c_spmat*>(&ref);
+}
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::SparseSolver(c_spmat *c_coefficients, c_spmat *c_observations)
+{
+  spmat row_major_coeffs = c_to_eigen(c_coefficients).pruned();
+  spmat row_major_observ = c_to_eigen(c_observations).pruned();
+
+  coefficients = new col_major_spmat(row_major_coeffs);
```

Comment:
I am a bit confused by this. Unless I am mistaken, we always use row major matrices in this module, but then we pretend that they are col major here. I understand this for symmetric or self-adjoint matrices, but many of the exposed solvers work on non-symmetric matrices.

Moreover, wouldn't we return a result in the wrong shape?

Reply:
The eigen solvers complained when I tried to pass in row-major matrices, so I added these conversions to make it work. I didn't realize this was just recasting the existing memory layout as column-major. My use case is symmetric, so I think I didn't notice that this is actually a bug. This sounds like it will be an issue. Do you know a way to run the solvers on row-major matrices?

Reply:
I suppose, in the absence of a way to call the solvers with row-major matrices we could (1) only expose those solvers that operate on symmetric matrices or (2) do a real memory re-map to column major format, though (2) might be too expensive to be worthwhile?

Reply:
I don’t know eigen well enough, I’ll look into it sometimes in the next days. @ryanrhymes you wrote the original bindings, maybe you know? Let’s wait a few days and see if we can find a workaround for this. If I don’t find it, I’ll ask for help on eigen repository. 

We could make a test with a rectangular matrix and see what happens if we want to double check my guess.

Reply:
I don't know it very well either honestly, sorry. Owl internally always uses row-major format, I originally suspected the conversion took too long but become unsure when Tyler said most of the time spent in solve function. I think an easier way is to run the same calculation in other libs like matab/python/julia so that we know how to benchmark.

Reply:
What does the solver compain actually? I am reading Eigen's doc, e.g. https://eigen.tuxfamily.org/dox/classEigen_1_1SparseQR.html Although the first parameter is matrix type, but the doc says it must col-major unfortunately. SparseLU has the same requirement, but not for SparseCholesky.

So I think the conversion from row-major to col-major is necessary for QR and LU.

Reply:
I began testing with LU and QR, I didn't try to user row-major matrices with other solvers after it failed with LU/QR. The error I saw was a compilation error saying "these functions may only be called with colum-major inputs."

Reply:
@steelet41 I think we only need an additional step in the templates.

According to the documentation (see https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html), ```
Matrices and arrays using one storage order can be assigned to matrices and arrays using the other storage order, as happens in the above program when Arowmajor is initialized using Acolmajor. Eigen will reorder the entries automatically. More generally, row-major and column-major matrices can be mixed in an expression as we want.
```

We could check if a matrix is symmetric, and in that case do nothing and proceed with the actual code, and otherwise have an intermediate cast to col-major as done in the documentation snippet, use that for the sparse solver and then cast the results back to row-major where needed.

Reply:
Okay, that's good news. I have actually settled on a different solution for my application (the python solver was just way faster...? a mystery I haven't solved) and I'm leaving the country for two weeks tomorrow, so I won't have a chance to work on this until next year at the earliest.

Reply:
There is no rush! Thank you very much. Is it the one in numpy/scipy another one?